File: src/wasm/native/lib/engine.cpp
#include <iostream>
#include <algorithm>
#include "./engine.h"
#include "./bin.h"
#include "./utils.h"
#include "./primitive.h"
#include "./box.h"

using namespace std;

vector<uint> Engine::getOrderedPrimIndices()
{
  return primitiveIndices;
}

vector<BvhNode> Engine::build(const vector<Primitive> &primitives)
{
  nodesUsed = 1;
  N = primitives.size();

  bvhNodes.resize(2 * N - 1);
  fill(bvhNodes.begin(), bvhNodes.end(), BvhNode());

  primitiveIndices.resize(N);
  for (uint i = 0; i < N; i++)
  {
    primitiveIndices[i] = i;
  }

  const uint rootNodeIdx = 0;
  BvhNode &root = bvhNodes[rootNodeIdx];
  root.leftFirst = 0;
  root.primCount = N;

  updateNodeBounds(rootNodeIdx, primitives);
  subdivide(rootNodeIdx, primitives);

  return bvhNodes;
}

void Engine::updateNodeBounds(const uint &nodeIdx, const vector<Primitive> &primitives)
{
  BvhNode &node = bvhNodes[nodeIdx];
  if (!node.primCount)
  {
    return;
  }

  uint last = node.leftFirst + node.primCount;
  for (uint i = node.leftFirst; i < last; i++)
  {
    const Primitive &primitive = primitives[primitiveIndices[i]];

    node.aabbMin.x = min(node.aabbMin.x, primitive.aabbMin.x);
    node.aabbMin.y = min(node.aabbMin.y, primitive.aabbMin.y);
    node.aabbMin.z = min(node.aabbMin.z, primitive.aabbMin.z);

    node.aabbMax.x = max(node.aabbMax.x, primitive.aabbMax.x);
    node.aabbMax.y = max(node.aabbMax.y, primitive.aabbMax.y);
    node.aabbMax.z = max(node.aabbMax.z, primitive.aabbMax.z);
  }
}

void Engine::subdivide(const uint &nodeIdx, const vector<Primitive> &primitives)
{
  BvhNode &node = bvhNodes[nodeIdx];

  BestSplit res = findBestSplitPlane(node, primitives);
  const unsigned short axis = res.axis;
  const float splitPlane = res.splitPlane;
  const float splitCost = res.splitCost;
  float nodeCost = calculateNodeCost(node);

  if (splitCost >= nodeCost)
  {
    return;
  }

  uint i = node.leftFirst;
  uint j = i + node.primCount - 1;
  while (i <= j)
  {
    const Primitive &primitive = primitives[primitiveIndices[i]];

    if (primitive.centroid[axis] < splitPlane)
    {
      i++;
    }
    else
    {
      swap(i, j--, primitiveIndices);
    }
  }

  uint leftCount = i - node.leftFirst;
  if (leftCount == 0 || leftCount == node.primCount)
  {
    return;
  }

  uint leftChildIdx = nodesUsed++;
  uint rightChildIdx = nodesUsed++;
  bvhNodes[leftChildIdx].leftFirst = node.leftFirst;
  bvhNodes[leftChildIdx].primCount = leftCount;
  bvhNodes[rightChildIdx].leftFirst = i;
  bvhNodes[rightChildIdx].primCount = node.primCount - leftCount;
  node.leftFirst = leftChildIdx;
  node.primCount = 0;
  updateNodeBounds(leftChildIdx, primitives);
  updateNodeBounds(rightChildIdx, primitives);
  subdivide(leftChildIdx, primitives);
  subdivide(rightChildIdx, primitives);
}

BestSplit Engine::findBestSplitPlane(const BvhNode &bvhNode, const vector<Primitive> &primitives)
{
  unsigned short bestAxis = 0;
  float bestPlane = -1;
  float bestCost = 1e30;
  const uint BIN_CNT = 4;
  const uint pCnt = bvhNode.primCount;

  for (unsigned short axis = 0; axis < 3; axis++)
  {
    float boundsMin = 1e30;
    float boundsMax = -1e30;

    for (uint i = 0; i < pCnt; i++)
    {
      const Primitive &primitive = primitives[primitiveIndices[bvhNode.leftFirst + i]];

      boundsMin = min(boundsMin, primitive.centroid[axis]);
      boundsMax = max(boundsMax, primitive.centroid[axis]);
    }

    if (boundsMax == boundsMin)
    {
      continue;
    }

    Bin bin[BIN_CNT];

    float rBinWidth = BIN_CNT / (boundsMax - boundsMin);
    for (uint i = 0; i < pCnt; i++)
    {
      const Primitive &primitive = primitives[primitiveIndices[bvhNode.leftFirst + i]];
      uint binIdx = min(BIN_CNT - 1, (uint)((primitive.centroid[axis] - boundsMin) * rBinWidth));
      bin[binIdx].primCount++;
      bin[binIdx].bounds.grow(primitive);
    }

    float leftArea[BIN_CNT - 1];
    float rightArea[BIN_CNT - 1];
    float leftCount[BIN_CNT - 1];
    float rightCount[BIN_CNT - 1];
    AABB leftBox, rightBox;
    float leftSum = 0, rightSum = 0;
    for (uint i = 0; i < BIN_CNT - 1; i++)
    {
      leftSum += bin[i].primCount;
      leftCount[i] = leftSum;
      leftBox.grow(bin[i].bounds);
      leftArea[i] = leftBox.area();
      rightSum += bin[BIN_CNT - 1 - i].primCount;
      rightCount[BIN_CNT - 2 - i] = rightSum;
      rightBox.grow(bin[BIN_CNT - 1 - i].bounds);
      rightArea[BIN_CNT - 2 - i] = rightBox.area();
    }

    float binWidth = 1 / rBinWidth;
    for (uint i = 0; i < BIN_CNT - 1; i++)
    {
      float cost = leftCount[i] * leftArea[i] + rightCount[i] * rightArea[i];

      if (cost < bestCost)
      {
        bestPlane = boundsMin + binWidth * (i + 1);
        bestAxis = axis;
        bestCost = cost;
      }
    }
  }

  return BestSplit{bestPlane, bestCost, bestAxis};
}

float Engine::calculateNodeCost(const BvhNode &bvhNode)
{
  return bvhNode.primCount * boxArea(bvhNode.aabbMin, bvhNode.aabbMax);
}

vector<uint> Engine::search(const Box &tgt)
{
  vector<uint> result;

  if (!bvhNodes.size())
  {
    return result;
  }

  intersect(tgt, 0, result);

  return result;
}

vector<vector<uint>> Engine::search(const vector<Box> &tgts)
{
  vector<vector<uint>> result;

  for (auto tgt : tgts)
  {
    result.push_back(search(tgt));
  }

  return result;
}

void Engine::intersect(const Box &tgt, const uint &nodeIdx, vector<uint> &result)
{
  if (!bvhNodes.size())
  {
    return;
  }

  BvhNode &node = bvhNodes[nodeIdx];
  if (!intersectAABB(tgt, node.aabbMin, node.aabbMax))
  {
    return;
  }

  if (node.isLeaf())
  {
    for (uint i = 0; i < node.primCount; i++)
    {
      result.push_back(primitiveIndices[node.leftFirst + i]);
    }
  }
  else
  {
    intersect(tgt, node.leftFirst, result);
    intersect(tgt, node.leftFirst + 1, result);
  }
}

bool Engine::intersectAABB(const Box &box, const float3 &bmin, const float3 &bmax)
{
  bool result = true;

  for (int axis = 0; axis < 3 && result; axis++)
  {
    result =
        result &&
        !(box.aabbMax[axis] < bmin[axis] || box.aabbMin[axis] > bmax[axis]);
  }

  return result;
}

File: src/js/Engine.ts
import { BVHNode } from './BVHNode';
import { AABB } from './AABB';
import { Bin } from './Bin';
import { BestSplit } from './Bestsplit';
import { expandToMin, expandToMax, surfaceArea, simpleDeepClone, calcAxialMidPoint } from './utils';
import type { IBoxALike } from './types';
import { search } from './searcher';

export class Engine {
  public static from(obj: {
    bvhNodes: BVHNode[],
    primitiveIndices: number[],
  }) {
    const e = new Engine();

    e.bvhNodes = simpleDeepClone(obj.bvhNodes);
    e.primitiveIndices = simpleDeepClone(obj.primitiveIndices);
    e.nodesUsed = e.bvhNodes!.length;

    return e;
  }

  private bvhNodes: BVHNode[] | null = null;
  private primitiveIndices: number[] | null = null;
  private nodesUsed: number = 1;
  private binCnt = 4;
  private lastBinIdx = this.binCnt - 1;
  private bin: Bin[];
  private leftArea: number[];
  private rightArea: number[];
  private leftCount: number[];
  private rightCount: number[];
  private leftBox: AABB;
  private rightBox: AABB;

  public get balanceFactor() {
    return this.binCnt;
  }

  constructor() {
    this.init();
  }

  private init() {
    this.lastBinIdx = this.binCnt - 1;

    this.bin = new Array<Bin>(this.binCnt);
    for (let i = 0; i < this.binCnt; i++) {
      this.bin[i] = new Bin();
    }

    this.leftArea = new Array<number>(this.lastBinIdx);
    this.rightArea = new Array<number>(this.lastBinIdx);
    this.leftCount = new Array<number>(this.lastBinIdx);
    this.rightCount = new Array<number>(this.lastBinIdx);
    this.leftBox = new AABB();
    this.rightBox = new AABB();
  }

  public build(primitives: IBoxALike[], balanceFactor?: number): void {
    if (balanceFactor && this.binCnt !== balanceFactor) {
      this.binCnt = balanceFactor;
      this.init();
    }

    this.nodesUsed = 1;

    const rootNodeIdx = 0;
    const N = primitives.length;

    const nodeTotal = 2 * N - 1;
    this.bvhNodes = new Array<BVHNode>(nodeTotal);
    for (let i = 0; i < nodeTotal; i++) {
      this.bvhNodes[i] = new BVHNode();
    }

    this.primitiveIndices = new Array<number>(N);
    for (let i = 0; i < N; i++) {
      this.primitiveIndices[i] = i;
    }

    const root = this.bvhNodes[rootNodeIdx];
    root.leftFirst = 0;
    root.primCount = N;

    this.updateNodeBounds(rootNodeIdx, primitives);
    this.subdivide(rootNodeIdx, primitives);
  }

  private updateNodeBounds(nodeIdx: number, primitives: IBoxALike[]): void {
    const node = this.bvhNodes![nodeIdx];

    if (!node.primCount) {
      return;
    }

    const last = node.leftFirst + node.primCount;
  
    for (let i = node.leftFirst; i < last; i++) {
      const primitive = primitives[this.primitiveIndices![i]];

      expandToMin(node.aabbMin, primitive.aabbMin);
      expandToMax(node.aabbMax, primitive.aabbMax);
    }
  }

  private subdivide(nodeIdx: number, primitives: IBoxALike[]): void {
    const node = this.bvhNodes![nodeIdx];
  
    const { axis, splitPlane, splitCost } = this.findBestSplitPlane(
      node,
      primitives
    );
    const nodeCost = this.calculateNodeCost(node);

    if (splitCost >= nodeCost) {
      return;
    }
  
    let i = node.leftFirst;
    let j = i + node.primCount - 1;
    while (i <= j) {
      const primitive = primitives[this.primitiveIndices![i]];
      const axialMidPoint = calcAxialMidPoint(primitive, axis);

      if (axialMidPoint < splitPlane) {
        i++;
      } else {
        this.swap(i, j--, this.primitiveIndices!);
      }
    }
  
    const leftCount = i - node.leftFirst;
    if (leftCount === 0 || leftCount === node.primCount) {
      return;
    }
  
    const leftChildIdx = this.nodesUsed++;
    const rightChildIdx = this.nodesUsed++;
    this.bvhNodes![leftChildIdx].leftFirst = node.leftFirst;
    this.bvhNodes![leftChildIdx].primCount = leftCount;
    this.bvhNodes![rightChildIdx].leftFirst = i;
    this.bvhNodes![rightChildIdx].primCount = node.primCount - leftCount;
    node.leftFirst = leftChildIdx;
    node.primCount = 0;
    this.updateNodeBounds(leftChildIdx, primitives);
    this.updateNodeBounds(rightChildIdx, primitives);
    this.subdivide(leftChildIdx, primitives);
    this.subdivide(rightChildIdx, primitives);
  }

  private findBestSplitPlane(bvhNode: BVHNode, primitives: IBoxALike[]): BestSplit {
    let bestAxis = 0;
    let bestPlane = 0;
    let bestCost = Infinity;
    const totalPrim = bvhNode.primCount;

    [0 /* x */, 1 /* y */, 2 /* z */].forEach((axis) => {
      this.bin.forEach(b => b.reset());
      this.leftBox.reset();
      this.rightBox.reset();

      let boundsMin = Infinity;
      let boundsMax = -Infinity;

      for (let i = 0; i < totalPrim; i++) {
        const primitive = primitives[this.primitiveIndices![bvhNode.leftFirst + i]];
        const axialMidPoint = calcAxialMidPoint(primitive, axis);

        boundsMin = Math.min(boundsMin, axialMidPoint);
        boundsMax = Math.max(boundsMax, axialMidPoint);
      }

      if (boundsMax === boundsMin) {
        return;
      }
  
      const rBinWidth = this.binCnt / (boundsMax - boundsMin);
      for (let i = 0; i < totalPrim; i++) {
        const primitive = primitives[this.primitiveIndices![bvhNode.leftFirst + i]];
        const axialMidPoint = calcAxialMidPoint(primitive, axis);
        const binIdx = Math.min(
          this.lastBinIdx,
          Math.floor((axialMidPoint - boundsMin) * rBinWidth)
        );
        this.bin[binIdx].primCount++;
        this.bin[binIdx].bounds.grow(primitive);
      }

      let leftSum = 0;
      let rightSum = 0;
      for (let i = 0; i < this.lastBinIdx; i++) {
        leftSum += this.bin[i].primCount;
        this.leftCount[i] = leftSum;
        this.leftBox.grow(this.bin[i].bounds);
        this.leftArea[i] = this.leftBox.area();
        rightSum += this.bin[this.lastBinIdx - i].primCount;
        this.rightCount[this.lastBinIdx - 1 - i] = rightSum;
        this.rightBox.grow(this.bin[this.lastBinIdx - i].bounds);
        this.rightArea[this.lastBinIdx - 1 - i] = this.rightBox.area();
      }
  
      const binWidth = (boundsMax - boundsMin) / this.binCnt;
      for (let i = 0; i < this.lastBinIdx; i++) {
        const cost = this.leftCount[i] * this.leftArea[i] + this.rightCount[i] * this.rightArea[i];

        if (cost < bestCost) {
          bestPlane = boundsMin + binWidth * (i + 1);
          bestAxis = axis;
          bestCost = cost;
        }
      }
    })
  
    return {
      axis: bestAxis,
      splitPlane: bestPlane,
      splitCost: bestCost,
    };
  }

  private calculateNodeCost(bvhNode: BVHNode): number {
    const area = surfaceArea(bvhNode);

    return bvhNode.primCount * area;
  }

  private swap(a: number, b: number, arr: number[]): void {
    [arr[a], arr[b]] = [arr[b], arr[a]];
  }

  public search(target: IBoxALike): number[] {
    const result: number[] = [];

    if (!this.bvhNodes?.length) {
      return result;
    }

    return search(target, {
      bvhNodes: this.bvhNodes,
      primitiveIndices: this.primitiveIndices
    });
  }
}
